src/language-js/needs-parens.js

```diff
@@ -16,37 +16,95 @@ function hasClosureCompilerTypeCastComment(text, path) {
   // Syntax example: var x = /** @type {string} */ (fruit);
 
   const n = path.getValue();
+  const parenStart = getParenStart(n);
 
   return (
-    isParenthesized(n) &&
-    (hasTypeCastComment(n) || hasAncestorTypeCastComment(0))
+    parenStart !== -1 &&
+    (getTypeCastIndex(n, parenStart, -1) !== -1 ||
+      hasAncestorTypeCastComment(0, parenStart))
   );
 
   // for sub-item: /** @type {array} */ (numberOrString).map(x => x);
-  function hasAncestorTypeCastComment(index) {
+  function hasAncestorTypeCastComment(index, parenStart) {
     const ancestor = path.getParentNode(index);
-    return ancestor && !isParenthesized(ancestor)
-      ? hasTypeCastComment(ancestor) || hasAncestorTypeCastComment(index + 1)
-      : false;
+    if (!ancestor) {
+      return false;
+    }
+
+    const ancestorParenStart = getParenStart(ancestor);
+    const typeCastIndex = getTypeCastIndex(
+      ancestor,
+      parenStart,
+      ancestorParenStart
+    );
+
+    // The node can inherit a typecast all the way to the root expression,
+    // provided some ancestor is not wrapped in parens. If so, then any
+    // typecast above that ancestor applies to the ancestor, not our node.
+    if (typeCastIndex === -1) {
+      return (
+        ancestorParenStart === -1 &&
+        hasAncestorTypeCastComment(index + 1, parenStart)
+      );
+    }
+
+    // There's an annoying issue if both our node contains a typecast, and this
+    // ancestor contains a typecast. Unfortunately, both comments are applied
+    // to ancestor. So we need to detect this situation, and insert hack paren
```

Comment:
Unfortunately, we can't rely on what node the comment will be attached to. It's not always the ancestor. E.g. see https://github.com/prettier/prettier/issues/7775.

Reply:
Ahh, I see. This is attaching two casts onto the same expression. I can fix this in a follow up PR using the same paren comment hack, if you'd like?